src/engine/source/kvdb/include/kvdb/kvdbHandlerCollection.hpp

```diff
@@ -0,0 +1,155 @@
+                               const std::string& dbName,
+                               const std::string& scopeName);
+
+    /**
+     * @brief Removes a KVDB Handler given the provided DB name and scope name.
+     * This automatically manage the reference counters and mappings.
+     *
+     * @param dbName Name of the DB.
+     * @param scopeName Name of the Scope.
+     *
+     */
+    void removeKVDBHandler(const std::string& dbName, const std::string& scopeName);
+
+    /**
+     * @brief Returns all the registered Database names.
+     *
+     * @return std::vector<std::string> List of DB names.
+     *
+     */
+    std::vector<std::string> getDBNames();
+
+    /**
+     * @brief Get the Reference Count of scopes for a given DB.
+     *
+     * @param dbName Name of the DB.
+     * @return std::map<std::string, int> Map of Scope names and their reference count.
+     */
+    std::map<std::string, int> getRefMap(const std::string& dbName);
+
+private:
+    /**
+     * @brief Helper class to manage the reference counters for a given DB.
+     * This is used to track how many scopes are using a given DB.
+     *
+     */
+    class KVDBHandlerInstance
+    {
+    public:
+        /**
+         * @brief Add a new scope to the reference counter.
+         *
+         * @param scopeName Name of the Scope.
+         *
+         */
+        void addScope(const std::string& scopeName);
+
+        /**
+         * @brief Remove a scope from the reference counter.
+         *
+         * @param scopeName Name of the Scope.
+         */
+        void removeScope(const std::string& scopeName);
+
+        /**
+         * @brief Returns if there are no scopes referencing this DB.
+         *
+         * @return true No scopes are referencing this DB.
+         * @return false Some scopes are referencing this DB.
+         */
+        bool emptyScopes() const;
+
+        /**
+         * @brief Get the list of scopes referencing this DB.
+         *
+         * @return std::vector<std::string> List of Scope names.
+         */
+        std::vector<std::string> getRefNames() const;
+
+        /**
+         * @brief Get the Ref Map object
+         *
+         * @return std::map<std::string, int> Map of DB names and their reference count.
+         */
+        std::map<std::string, int> getRefMap() const;
+
+    private:
+        /**
+         * @brief Basically a wrapper to std::map<std::string, int> with some helper functions.
+         *
+         */
+        RefCounter m_scopeCounter;
+    };
+
+private:
+    /**
+     * @brief Internal map holding DB names and their KVDBHandlerInstance with extra info.
+     *
+     */
+    std::map<std::string, std::shared_ptr<KVDBHandlerInstance>> m_mapInstances;
+
+    /**
+     * @brief Pointer to the Manager that deals with handlers.
+     *
+     */
+    IKVDBHandlerManager* m_handleManager {nullptr};
+
+    /**
+     * @brief Mutex to protect the internal map.
+     *
+     */
+    std::mutex m_mutex;
```

Comment:
Is it necessary to do mutual exclusion? perhaps, if we mostly read and not write the map, it is convenient to use std::shared_mutex. 
If the mutex is only needed at compile time of the execution pipeline (environment construction) and not at run time of the expression chain, it would not be necessary. What do you think?